apps/manager_app/firmware_update.c

```diff
@@ -0,0 +1,195 @@
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static bool check_which_request(const manager_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->firmware_update.which_request) {
+    manager_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                       ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_query(manager_firmware_update_request_t *request) {
+  if (false == request->initiate.has_version) {
+    manager_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                       ERROR_DATA_FLOW_FIELD_MISSING);
+    return false;
+  }
+  uint32_t current_version = get_fwVer();
+
+  common_version_t *target = &request->initiate.version;
+  uint32_t target_version =
+      (target->major << 24) | (target->minor << 16) | (target->patch);
+
+  // Query is invalid if the target version is equal or less than the current
+  // firmware version installed
+  if (target_version <= current_version) {
+    manager_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                       ERROR_DATA_FLOW_INVALID_DATA);
+    return false;
+  }
+
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void manager_confirm_firmware_update(manager_query_t *query) {
+  if ((!check_which_request(
+          query, MANAGER_FIRMWARE_UPDATE_INITIATE_REQUEST_VERSION_TAG)) ||
+      (!validate_query(&query->firmware_update))) {
+    return;
+  }
+
+  char msg[100];
+  snprintf(msg,
+           sizeof(msg),
+           FIRMWARE_UPDATE_CONFIRMATION,
+           query->firmware_update.initiate.version.major,
+           query->firmware_update.initiate.version.minor,
+           query->firmware_update.initiate.version.patch);
+  if (!core_user_confirmation(
+          NULL, msg, CONFIRMATION_SCREEN, manager_send_error)) {
+    return;
+  }
+
+  manager_result_t result =
+      init_manager_result(MANAGER_RESULT_FIRMWARE_UPDATE_TAG);
+  manager_firmware_update_response_t *resp = &result.firmware_update;
+  resp->which_response = MANAGER_FIRMWARE_UPDATE_RESPONSE_CONFIRMED_TAG;
+  resp->confirmed.dummy_field = '\0';
+  manager_send_result(&result);
+
+  // NOTE: This is a USB initiated flow, however, device will go in bootloader
+  // mode after blocking delay of 500ms without serving any events. So in case
+  // any abort event is triggered by the host, it will NOT be served!
+  // Wait for status pull to desktop (which requests at 200ms)
+  instruction_scr_init(ui_text_processing, NULL);
+  BSP_DelayMs(500);
```

Comment:
can add following comment.

`// hard delay to save from glitchy behaviour`

Reply:
Hey, while you are right, but this delay is not to prevent glitchy behaviour, we could have used delay screen for that.

Delay is to ensure that user confirmation response is polled by the host which is done periodically every 200ms.